File: packages/pvm-assemblyscript/assembly/host-functions/general/lookup.ts
import { HostFunctionResult } from '../accumulate/base'
import {
  HostFunctionContext,
  HostFunctionParams,
  LookupParams,
} from './base'
import { CompleteServiceAccount } from '../../codec'
import {
  ACCUMULATE_ERROR_CODES,
  GENERAL_FUNCTIONS,
  RESULT_CODES,
} from '../../config'
import { BaseHostFunction } from './base'

/**
 * LOOKUP host function (Ω_L)
 *
 * Looks up preimages from service account storage
 *
 * Gray Paper Specification (pvm_invocations.tex lines 374-396):
 * - Function ID: 2 (lookup)
 * - Gas Cost: 10
 * - Signature: Ω_L(gascounter, registers, memory, s, s, d)
 *   - s = current service account
 *   - s = current service ID
 *   - d = accounts dictionary
 *
 * Service account selection (a):
 *   a = s (current service)       when registers[7] ∈ {s, 2^64 - 1}
 *   a = d[registers[7]]           when registers[7] ∈ keys{d}
 *   a = none                      otherwise
 *
 * Register usage:
 *   registers[7] = service ID to query (or s/2^64-1 for self)
 *   registers[8] = hash offset (h) - 32 bytes to read from memory
 *   registers[9] = output offset (o) - where to write result
 *   registers[10] = from offset (f) in preimage
 *   registers[11] = length (l) to copy
 *
 * Value lookup (v):
 *   v = error  when hash memory not readable
 *   v = none   when a = none OR memory[h:32] ∉ keys{a.preimages}
 *   v = a.preimages[memory[h:32]] otherwise
 *
 * Slice parameters:
 *   f = min(registers[10], len{v})
 *   l = min(registers[11], len{v} - f)
 *
 * Result:
 *   - PANIC if v = error OR output memory not writable
 *   - registers[7] = NONE if v = none
 *   - registers[7] = len{v}, memory[o:l] = v[f:l] otherwise
 *
 * NOTE: LOOKUP is READ-ONLY - it does NOT modify service accounts!
 */

export class LookupHostFunction extends BaseHostFunction {
  functionId: u64 = GENERAL_FUNCTIONS.LOOKUP
  name: string = 'lookup'

  execute(
    context: HostFunctionContext,
    params: HostFunctionParams | null,
  ): HostFunctionResult {
    if (!params) {
      return new HostFunctionResult(RESULT_CODES.PANIC)
    }
    const lookupParams = params as LookupParams
    const queryServiceId = context.registers[7]
    const hashOffset = context.registers[8]
    const outputOffset = context.registers[9]
    const fromOffset = context.registers[10]
    const length = context.registers[11]

    // Gray Paper: a = service account to query
    // a = s (self)       when registers_7 ∈ {s, 2^64 - 1}
    // a = d[registers_7] when registers_7 ∈ keys{d}
    // a = none           otherwise
    const MAX_U64: u64 = u64.MAX_VALUE // 2^64 - 1

    let serviceAccount: CompleteServiceAccount | null = null

    if (
      queryServiceId == lookupParams.serviceId ||
      queryServiceId == MAX_U64
    ) {
      // Query self
      if (lookupParams.accounts.has(lookupParams.serviceId)) {
        serviceAccount = lookupParams.accounts.get(lookupParams.serviceId)
      }
    } else if (lookupParams.accounts.has(queryServiceId)) {
      // Query another service from accounts dictionary
      serviceAccount = lookupParams.accounts.get(queryServiceId)
    }
    // else: serviceAccount remains null (a = none)

    // Read hash from memory (32 bytes)
    // Gray Paper: v = error when N[h,32] ⊄ readable{memory}
    const readResult_hashData = context.ram.readOctets(u32(hashOffset), 32)
    const hashData = readResult_hashData.data
    const hashFaultAddress = readResult_hashData.faultAddress
    if (hashData === null || hashFaultAddress !== 0) {
      // v = error - memory not readable
      return new HostFunctionResult(RESULT_CODES.PANIC)
    }

    // Gray Paper: v = none when a = none ∨ memory[h:32] ∉ keys{a.preimages}
    if (serviceAccount === null) {
      // a = none - service account not found
      context.registers[7] = ACCUMULATE_ERROR_CODES.NONE
      return new HostFunctionResult(255) // continue execution
    }

    // Look up preimage by hash
    const preimage = serviceAccount.preimages.get(hashData)
    if (!preimage) {
      // v = none - preimage not found
      context.registers[7] = ACCUMULATE_ERROR_CODES.NONE
      return new HostFunctionResult(255) // continue execution
    }

    // v found - calculate slice parameters
    // Gray Paper: f = min(registers[10], len{v})
    //             l = min(registers[11], len{v} - f)
    const preimageLength = preimage.length
    const f: i32 = fromOffset < u64(preimageLength)
      ? i32(fromOffset)
      : preimageLength
    const remainingAfterF: i32 = preimageLength - f
    const l: i32 = length < u64(remainingAfterF)
      ? i32(length)
      : remainingAfterF

    // Only write if there's data to write
    if (l > 0) {
    // Extract data slice
      const dataToWrite = preimage.slice(f, f + l)

    // Write preimage slice to memory
      // Gray Paper: PANIC if N[o,l] ⊄ writable{memory}
    const writeResult = context.ram.writeOctets(u32(outputOffset), dataToWrite)
    if (writeResult.hasFault) {
      return new HostFunctionResult(RESULT_CODES.PANIC)
      }
    }

    // Gray Paper: Return len{v} (full preimage length, not slice length)
    context.registers[7] = u64(preimageLength)

    return new HostFunctionResult(255) // continue execution
  }
}

File: packages/pvm-rust/src/host_functions/general/lookup.rs
//! LOOKUP host function (Ω_L). Gray Paper: function ID 2.
//! Preimage lookup by hash; r7=serviceId, r8=hash offset, r9=output offset, r10=from, r11=length.
//! 1:1 with AS: no params → r7=NONE; with params resolve account, get preimage, write slice, r7=len.

use crate::codec::{get_preimage_value, CompleteServiceAccount};
use crate::config::{FUNC_LOOKUP, REG_NONE};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

const MAX_U64: u64 = u64::MAX;

/// LOOKUP (2): read hash from memory; no accounts → r7 = NONE; else resolve account, preimage lookup, write slice.
pub struct LookupHostFunction;

impl HostFunction for LookupHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_LOOKUP
    }
    fn name(&self) -> &'static str {
        "lookup"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let query_service_id = context.registers[7];
        let hash_offset = context.registers[8];
        let output_offset = context.registers[9];
        let from_offset = context.registers[10];
        let length = context.registers[11];

        let read_hash = context.ram.read_octets(hash_offset as u32, 32);
        if read_hash.data.is_none() || read_hash.fault_address != 0 {
            crate::host_log!(
                "[hostfn] lookup PANIC: hash read fault (offset={}, fault_address={})",
                hash_offset, read_hash.fault_address
            );
            return HostFunctionResult::panic();
        }
        let hash_data = read_hash.data.as_ref().unwrap();

        let (service_account, actual_service_id): (Option<&CompleteServiceAccount>, u32) = match (&context.service_id, &context.accounts) {
            (Some(service_id), Some(accounts)) => {
                let sid_val = *service_id;
                let account = if query_service_id == sid_val || query_service_id == MAX_U64 {
                    accounts.get(&sid_val)
                } else {
                    accounts.get(&query_service_id)
                };
                match account {
                    Some(a) => {
                        let sid = if query_service_id == MAX_U64 {
                            sid_val
                        } else {
                            query_service_id
                        };
                        (Some(a), sid as u32)
                    }
                    None => {
                        crate::host_log_error!(
                            "[hostfn] lookup: Service account not found (queryServiceId={})",
                            query_service_id
                        );
                        context.registers[7] = REG_NONE;
                        return HostFunctionResult::continue_execution();
                    }
                }
            }
            _ => {
                context.registers[7] = REG_NONE;
                return HostFunctionResult::continue_execution();
            }
        };

        let service_account = match service_account {
            Some(a) => a,
            None => {
                context.registers[7] = REG_NONE;
                return HostFunctionResult::continue_execution();
            }
        };
        let preimage = get_preimage_value(service_account, actual_service_id, hash_data);
        let preimage = match preimage {
            Some(p) => p,
            None => {
                crate::host_log_error!(
                    "[hostfn] lookup: Preimage not found (serviceId={}, hashOffset={})",
                    actual_service_id, hash_offset
                );
                context.registers[7] = REG_NONE;
                return HostFunctionResult::continue_execution();
            }
        };

        let preimage_len = preimage.len() as i32;
        let f = from_offset.min(preimage_len as u64) as i32;
        let remaining = (preimage_len - f).max(0);
        let l = (length as i64).min(remaining as i64).max(0) as usize;

        if l > 0 {
            let end = (f as usize + l).min(preimage.len());
            let data_to_write = preimage[f as usize..end].to_vec();
            let write_result = context.ram.write_octets(output_offset as u32, &data_to_write);
            if write_result.has_fault {
                crate::host_log!(
                    "[hostfn] lookup PANIC: output write fault (offset={}, len={})",
                    output_offset, data_to_write.len()
                );
                return HostFunctionResult::panic();
            }
        }

        context.registers[7] = preimage_len as u64;
        HostFunctionResult::continue_execution()
    }
}
